lib/vrl/compiler/src/shared_value/target.rs

```diff
@@ -0,0 +1,560 @@
+                .borrow()
+                .as_object()
+                .and_then(|map| fields.iter().find_map(|field| map.get(field.as_str())))
+                .cloned(),
+            SegmentBuf::Index(index) => self.borrow().as_array().and_then(|array| {
+                let len = array.len() as isize;
+                if *index >= len || index.abs() > len {
+                    return None;
+                }
+
+                index
+                    .checked_rem_euclid(len)
+                    .and_then(|i| array.get(i as usize))
+                    .cloned()
+            }),
+        }
+    }
+
+    /*
+     * We don't need _mut anymore sinc Rc<RefCell> is runtime mutable.
+    fn get_by_segments_mut<'a, T>(&mut self, mut segments: T) -> Option<Rc<RefCell<Value>>>
+    where
+        T: Iterator<Item = &'a SegmentBuf>,
+    {
+        let segment = match segments.next() {
+            Some(segments) => segments,
+            None => return Some(self),
+        };
+
+        self.get_by_segment_mut(segment)
+            .and_then(|value| value.borrow_mut().get_by_segments_mut(segments).clone())
+    }
+
+    fn get_by_segment_mut(&mut self, segment: &SegmentBuf) -> Option<Rc<RefCell<Value>>> {
+        match segment {
+            SegmentBuf::Field(FieldBuf { name, .. }) => self
+                .as_object_mut()
+                .and_then(|map| map.get_mut(name.as_str()))
+                .cloned(),
+            SegmentBuf::Coalesce(fields) => self.as_object_mut().and_then(|map| {
+                fields
+                    .iter()
+                    .find(|field| map.contains_key(field.as_str()))
+                    .and_then(move |field| map.get_mut(field.as_str()))
+                    .cloned()
+            }),
+            SegmentBuf::Index(index) => self.as_array_mut().and_then(|array| {
+                let len = array.len() as isize;
+                if *index >= len || index.abs() > len {
+                    return None;
+                }
+
+                index
+                    .checked_rem_euclid(len)
+                    .and_then(move |i| array.get_mut(i as usize))
+                    .cloned()
+            }),
+        }
+    }
+    */
+
+    fn remove_by_segments<'a, T>(self, mut segments: Peekable<T>, compact: bool)
+    where
+        T: Iterator<Item = &'a SegmentBuf> + Clone,
+    {
+        let segment = match segments.next() {
+            Some(segments) => segments,
+            None => {
+                return match &mut *self.borrow_mut() {
+                    Value::Object(v) => v.clear(),
+                    Value::Array(v) => v.clear(),
+                    _ => {
+                        // TODO This needs serious testing.
+                        self.replace(Value::Null);
+                    }
+                };
+            }
+        };
+
+        if segments.peek().is_none() {
+            return self.borrow_mut().remove_by_segment(segment);
```

Comment:
The code here is probably the hairiest aspect of this PR. There are a lot of `borrow` and `borrow_mut` calls around.

With a bit of work it's possible a lot of this could be tidied up so we only have a single `borrow_mut` at the top, but I think it would still be of great value to fuzz test this code very thoroughly.